### DISCUSSION TITLE
Generic shell API

### DISCUSSION THREAD

**Discussion by elviskahoro** [warpdotdev/Warp#641] [Product roadmap]
Generic shell API

**Comment by elviskahoro** [comment]
### Describe the solution you'd like?

```shell
I noticed that quick move of a cursor does not work if I enter a docker container. When I do docker exec -it ... bash cmd + arrow (and option + arrow) combination stops working.
```


### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

_No response_

**Reply by elviskahoro** [reply]
Docker

From @SMores via #379

I'd like to add to this a bit:

It would be _lovely_ if `docker exec` could get the same treatment as `ssh`. I do the vast majority of my development within docker containers, which means most of my terminal usage is within docker containers (mostly via VS Code Development Containers). It would be really awesome to be able to use Warp for that! But right now, it falls back to a somewhat rough basic terminal emulation instead, with broken up and down arrow keys (they navigate Warp Blocks instead of bash/zsh history) as @elviskahoro mentioned

**Reply by MunsMan** [reply]
Unsure how to understand this Discussion, but are there any updates about a docker exec? I would love this feature and are willing to support if needed.

**Comment by elviskahoro** [comment]
Support pgcli

**Comment by elviskahoro** [comment]
Support kubie

**Comment by elviskahoro** [comment]
psql: can not move the caret with cmd key

From @oguzhantopcu via #586

### Describe the bug

```shell
i can not move my caret with [command] key on "psql", "git add -i" CLI's.

*psql is postgresql's official cli.

it makes psql pretty unusable.
```


### To Reproduce

Steps to reproduce:

1.) run psql.
2.) type "something"
3.) press [command]+[left]

caret won't move to the left side of the word, instead it types [space].

### Expected behaviour

```shell
caret should move to the left side of the word
```


### Screenshots

_No response_

### Operating System

MacOS

### OS Version

12.1

### Additional context

_No response_

**Comment by elviskahoro** [comment]
Support nu shell?

From @WindSoilder via #613

### Describe the solution you'd like?

```shell
As title, I wonder if nu shell(https://github.com/nushell/nushell) can be supported?
```


### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

_No response_

**Reply by KieranP** [reply]
Warp + NuShell would be an excellent combo for modern day

**Reply by MitchyCola** [reply]
I tried Warp with nushell on a whim, the blocks do not work. Would love nushell support.

**Reply by evelant** [reply]
Warp looks cool but I'm all in on nushell. I just can't go back to the tangled mess of syntax and quirks that is old shells.

**Comment by elviskahoro** [comment]
From amirali7e via Discord:
https://discord.com/channels/851854972600451112/906269689116770304/935234587347550248
> Have been using warp for a while, pretty cool product you have developed. 
> Are there any plans to support REPLs ? Particularly Julia which would greatly benefit from such a product since it has a heavily REPL based development. I will bring "warp" up in the community(better yet you can post in their discourse), they are probably more than willing to help.

**Reply by elviskahoro** [reply]
> http://discourse.julialang.org/

**Comment by elviskahoro** [comment]
From [taranlu-houzz](https://github.com/taranlu-houzz) via #858

### Describe the solution you'd like?

```text
It would be great to also have support for `xonsh` shell: https://xon.sh/

I primarily use `fish`, but often dip into `xonsh` to work on specific problems interactively with Python. It would be nice to have first class support, so that commands show up in blocks, etc.
```


### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

_No response_

**Comment by siracusa** [comment]
I use tcsh as my login shell and I'd love to see it supported by Warp in addition to Bash, Zsh, and Fish.

**Comment by jnoortheen** [comment]
> Generic shell API

This can be done by using shell integration . Most of the modern terminanls follow the iTerm's OSC tokens for marking the boundaries.
   - https://sw.kovidgoyal.net/kitty/shell-integration/
  
What I really love to see is the autocompletion support from shell as well as user defined code. Something like https://github.com/withfig/autocomplete

**Comment by darthwalsh** [comment]
To add support for another shell (i.e. I'd like `pwsh` from #880) is there a protocol spec for how Warp Terminal communicates with the shell? I don't know what `preexec` is (mentioned on https://www.warp.dev/faq) but I guess that's part of the protocol.

**Reply by jtsalva** [reply]
+1 would like to see `pwsh` support

**Reply by sgiddings-kunzleigh** [reply]
I would absolutely love PowerShell support as well as support for Nushell. But as I work most of my day inside of Azure and Microsoft 365, it would be awesome to have the same Zsh style of support for PowerShell.

**Comment by warpdotdev-devx** [comment]
We don't have an API or endpoint yet for this.

--- original message ---
On April 14, 2022, 8:55 PM EDT ***@***.*** wrote:



To add support for another shell (i.e. I'd like pwsh from #880) is there a protocol spec for how Warp Terminal communicates with the shell? I don't know what preexec is (mentioned on https://www.warp.dev/faq) but I guess that's part of the protocol.

—

Reply to this email directly, view it on GitHub, or unsubscribe.

You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

-- 
You received this message because you are subscribed to the Google Groups "github" group.
To unsubscribe from this group and stop receiving emails from it, send an email to ***@***.***
--- end of original message ---

**Reply by BatmanAoD** [reply]
Isn't the concept of "Device Control Strings" itself an API? Why can't shells that want to support Warp blocks just have an option to emit DCSes?

**Comment by OmerFlame** [comment]
It would be a crazy improvement in everyone's productivity if you guys add `xonsh` support, it's a shell that can uses Python scripts instead of shell with its own custom interpreter for adding shell-like functionality. This sounds like a crazy good idea and I would absolutely love to see support for it in Warp!

**Comment by dertuxmalwieder** [comment]
I would love to have an easy way to write my own shell glue code. I tend to use rather obscure shells like `rc` and I wouldn't expect the Warp team to know and support all of them.
